LoongArch-ELF-ABI-EN.adoc

```diff
@@ -0,0 +1,206 @@
+= LoongArch ELF ABI specification
+Loongson Technology Corporation Limited
+V1.00
+
+:encoding: UTF-8
+:frame: top
+:grid: rows
+:valign: middle
+
+== Register Convention
+.Integer Register Convention
+[width="80%",options="header",cols="2m,2m,^5,^3"]
+|====================================================================
+|Name     |Alias        |Meaning             |Preserved across calls
+|$r0      |$zero        |Constant zero       |--
+|$r1      |$ra          |Return address      |No
+|$r2      |$tp          |Thread pointer      |--(Unallocatable)
+|$r3      |$sp          |Stack pointer       |Yes
+|$r4-$r11 |$a0-$a7      |Argument registers  |No
+|$r4-$r5  |$v0-$v1      |Return value        |No
+|$r12-$r20|$t0-$t8      |Temp registers      |No
+|$r21     |$x           |Reserved            |--(Unallocatable)
+|$r22     |$fp          |Frame pointer       |Yes
+|$r23-$r31|$s0-$s8      |Subroutine register variables   |Yes
+|====================================================================
+
+.Floating-point Register Convention
+[width="80%",options="header",cols="2m,2m,^5,^3"]
+|====================================================================
+|Name     |Alias        |Meaning            |Preserved across calls
+|$f0-$f7  |$fa0-$fa7    |Argument registers |No
+|$f0-$f1  |$fv0-$fv7    |Return value       |No
+|$f8-$f23 |$ft0-$ft15   |Temp registers     |No
+|$f24-$f31|$fs0-$fs7    |Subroutine register variables   |Yes
+|====================================================================
+
+
+== Type Size and Alignment
+
+.LP64 ABI
+[width="80%",options="header",cols="2m,^5,^3"]
+|====================================================================
+|Scalar type               |Size |Alignment
+|unsigned/signed char      |1    |1
+|unsigned/signed short     |2    |2
+|unsigned/signed int       |4    |4
+|unsigned/signed long      |8    |8
+|unsigned/signed long long |8    |8
+|pointer                   |8    |8
+|float                     |4    |4
+|double                    |8    |8
+|long double               |16   |16
+|====================================================================
+
+== ELF Object Files
+
+e_machine: LoongArch (258)
+
+e_flags: LP32 (0x1) LP64 (0x3)
```

Comment:
No soft-float considerations?

Reply:
Meaning?

If you just want to distinguish 32-bit from 64-bit, use `EI_CLASS`

Reply:
Good idea, more ABIs should be planed and listed.

Reply:
Also, according to [some][src1] [sources][src2], the 32-bit-only ABI should perhaps be named "ILP32" instead of "LP32".

[src1]: https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models
[src2]: https://nickdesaulniers.github.io/blog/2016/05/30/data-models-and-word-size/